### PR TITLE
FOGL-2947: Pi_Server plugin : handle kerberos authentication using the libcurl library

### PR DESCRIPTION
**Pull request by sstefano69**
*Working stage*  

**IMPORTANT NOTE**
we use the curl repo : 
https://github.com/curl/curl/releases/download/curl-7_65_3

tested/to be tested : 
* 1st test on a fresh ubuntu 18
* centos fresh - to be tested
* raspbian fresh - to be tested

```
to install/cfg kerberos are need, done by requirements.sh  - ubuntu :
sudo apt-get install -y krb5-user
sudo apt-get install -y libcurl4-openssl-dev

answers : 
* DIANOMIC.COM
* win-4m7odkb0rh2.dianomic.com
* win-4m7odkb0rh2.dianomic.com

sudo vi  /etc/hosts
192.168.1.51    win-4m7odkb0rh2.dianomic.com win-4m7odkb0rh2

#copy the keybtab file in 
${FOGLAMP_DATA}/data/etc/kerberos/piwebapi_kerberos_https.keytab
```

Configuration : 
image

erro handling fixed : 
```
Aug 13 11:23:40 rdev2 North_Readings_to_PI[6842]: ERROR: Sending JSON data error: |HTTP code |500| - HTTP error |HTTP/1.1 500 Internal Server Error|| - HostPort |WIN-4M7ODKB0RH2:443| - path |/piwebapi/omf| - OMF message |[{"containerid": "
```

added check : 
```
Aug 13 13:50:51 rzero North_Readings_to_PI[15755]: ERROR: Kerberos authentication not possible, the keytab file :/home/foglamp/Development/FogLAMP/data/etc/kerberos/piwebapi_kerberos_https.keytab: is missing.
```

### CODE REVIEW

File: C/plugins/common/libcurl_https.cpp [hunk 1]
```diff
@@ -66,15 +54,53 @@ LibcurlHttps::LibcurlHttps(const string& host_port,
...
-	/* free the custom headers */
-	//curl_slist_free_all(chunk);
+// Handles payload using the libcurl structures
+struct WriteThis {
+    const char *readptr;
+    size_t      sizeleft;
+};
 
-	curl_global_cleanup();
+// Callback for libcurl
```
Comment (pintomax) [inline C/plugins/common/libcurl_https.cpp]: Add description for input parameters

**Comment by Singhal-Vaibhav** [comment]
@sstefano69  Installed the desired package for this pull on PR tester machine. The krb5-user library asked for other kerberos configuration which are currently set to default. If they should'nt be set to default, please let me know.  Please rebase your pull.

**Comment by sstefano69** [comment]
@Singhal-Vaibhav the answers I used so fare are : 
```
answers : 
* DIANOMIC.COM
* win-4m7odkb0rh2.dianomic.com
* win-4m7odkb0rh2.dianomic.com
```
updating the host  : 
```
sudo vi  /etc/hosts
192.168.1.51    win-4m7odkb0rh2.dianomic.com win-4m7odkb0rh2
```
but the should not be relevant for the build only.

**Review by MarkRiddoch — DISMISSED** [review]
Looks fine to me

File: C/common/string_utils.cpp [hunk 2]
```diff
@@ -34,3 +34,25 @@ void StringReplace(std::string& StringToManage,
 				       StringReplacement);
 	}
 }
+
+/**
+ * Strips Line feed and carige return
+ *
+ */
+void StringStripCRLF(std::string& StringToManage)
```
Comment (AmandeepArora) [inline C/common/string_utils.cpp:42]: I guess only a single CR & LF characters are expected in input string?

**Comment by AmandeepArora** [comment]
Just a minor clarification. Looks good otherwise.

**Review by pintomax — DISMISSED** [review]

**Comment by sstefano69** [comment]
> If the requirements.sh script is already run by root user, the sudo before make install could be removed.
> Might be worth checking the Jenkins jobs

solved.

File: C/common/string_utils.cpp (hunk 2 above)
Comment (sstefano69) [reply C/common/string_utils.cpp:42]: > I guess only a single CR & LF characters are expected in input string?

we had to manage the HTTP headers the obtain a description from a HTTP request.
the headers contain CR & LF and so we used this function to strip them.

**Review by pintomax — DISMISSED** [review]

**Review by MarkRiddoch — APPROVED** [review]

**Pull request merged**